pkg/registry/pod/strategy.go

```diff
@@ -65,6 +65,21 @@ func (podStrategy) PrepareForUpdate(obj, old runtime.Object) {
 	newPod := obj.(*api.Pod)
 	oldPod := old.(*api.Pod)
 	newPod.Status = oldPod.Status
+	// TODO: VolumeMounts of init containers currently aren't merged properly
+	// by the PATCH handler because they are stored as a string under
+	// PodInitContainersAnnotationKey (pod.alpha.kubernetes.io/init-containers)
+	// in annotations of v1.PodSpec.
+	// In the existing Pod, VolumeMounts of init containers may be altered by
+	// secrets / config maps mechanisms causing ValidatePodUpdate to fail.
+	// As a workaround, changes in VolumeMounts of init containers are
+	// currently ignored and don't cause validation errors.
+	// Remove this workaround when InitContainers go out of alpha and
+	// PodInitContainersAnnotationKey is removed.
+	if len(oldPod.Spec.InitContainers) == len(newPod.Spec.InitContainers) {
```

Comment:
InitContainers are alpha, and at this point I would prefer to not fix this
in alpha.  When we move it to beta, this problem goes away.

This is a general acceptable limitation of alpha resources.

On Thu, Jul 14, 2016 at 11:37 AM, Ivan Shvedunov <notifications@github.com>
wrote:

> In pkg/registry/pod/strategy.go
> <https://github.com/kubernetes/kubernetes/pull/28557#discussion_r70828383>
> :
>
> > @@ -65,6 +65,21 @@ func (podStrategy) PrepareForUpdate(obj, old runtime.Object) {
> >  	newPod := obj.(*api.Pod)
> >  	oldPod := old.(*api.Pod)
> >  	newPod.Status = oldPod.Status
> > +	// TODO: VolumeMounts of init containers currently aren't merged properly
> > +	// by the PATCH handler because they are stored as a string under
> > +	// PodInitContainersAnnotationKey (pod.alpha.kubernetes.io/init-containers)
> > +	// in annotations of v1.PodSpec.
> > +	// In the existing Pod, VolumeMounts of init containers may be altered by
> > +	// secrets / config maps mechanisms causing ValidatePodUpdate to fail.
> > +	// As a workaround, changes in VolumeMounts of init containers are
> > +	// currently ignored and don't cause validation errors.
> > +	// Remove this workaround when InitContainers go out of alpha and
> > +	// PodInitContainersAnnotationKey is removed.
> > +	if len(oldPod.Spec.InitContainers) == len(newPod.Spec.InitContainers) {
>
> You may try removing this part and the new e2e test added in this PR will
> break although it doesn't change anything at all, just applying the same
> yaml
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28557/files/633baf053b8a7fd7862d5881b048bcf1b5a9624e#r70828383>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p4HmU0sKIv2TQaWUtOPLSCAXzKM0ks5qVlesgaJpZM4JGkjJ>
> .
>